README.md
This is a simple spec-compliant GraphQL rack application and middleware based on the [graphql](https://github.com/rmosolgo/graphql-ruby) gem. Since it's built with Rack, it can be mounted with most ruby web servers.

Install the gem:

```
gem install graphql_server
```

# Using the server

## As a standalone application

```ruby
# app.ru
require 'graphql_server'

type_def = <<-GRAPHQL
  type Query {
    hello: String
  }
GRAPHQL

resolver = {
  "Query" => {
    "hello" => Proc.new { "world" }
  }
}

run GraphQLServer.new(type_def: type_def, resolver: resolver)
```

Start using `rackup`

```
rackup app.ru
```

## As a middleware in your existing application

```ruby
# app.ru
require 'graphql_server'

type_def = ...
resolver = ...

use GraphQLServer, type_def: type_def, resolver: resolver
```

# Options

## Schema

You can get started fast by writing a type defintions and a resolver hash

```ruby
GraphQLServer.new(type_def: type_def, resolver: resolver)
```

You can also provide your own [schema](http://graphql-ruby.org/schema/definition.html)

```ruby
GraphQLServer.new(schema: schema)
```

See the examples folder for more details

# GraphQL Playground

You can use the excellent GraphQL Playground IDE from Prisma

```
gem install graphql_playground
```

Map it to the url of your choice and point it to your GraphQL server endpoit

```ruby
map '/playground' do
  use GraphQLPlayground, endpoint: '/'
end

run GraphQLServer.new(type_def: type_def, resolver: resolver)
```


CHANGELOG.md
# Changelog

## v0.1.1

- Set Ruby 2.2.0 as the required ruby version in gemspec

## v0.1.0

- Changed `GraphqlServer` class name to `GraphQLServer` to follow the graphql gem convention
- Removed `path` as an option for middleware since it can be mapped with rack
- Fix a few small issues with the code


test/hello_world_server.rb
class HelloWorldServer
  class << self
    def app
      type_def = <<-GRAPHQL
        type Query {
          hello: String
        }
      GRAPHQL
      
      resolver = {
        "Query" => { 
          "hello" => Proc.new { "world" }
        }
      }

      Rack::Builder.new do
        run GraphQLServer.new(type_def: type_def, resolver: resolver)
      end
    end

    def middleware(path = nil)
      type_def = <<-GRAPHQL
        type Query {
          hello: String
        }
      GRAPHQL
      
      resolver = {
        "Query" => { 
          "hello" => Proc.new { "world" }
        }
      }

      Rack::Builder.new do
        use GraphQLServer, type_def: type_def, resolver: resolver
        run [200]
      end
    end
  end
end

test/graphql_server_test.rb
require 'minitest/autorun'
require "minitest/spec"

require 'graphql_server'
require 'rack/test'
require 'hello_world_server'

describe GraphQLServer do
  include Rack::Test::Methods

  describe "app" do
    describe "#call" do
      let(:app) { HelloWorldServer.app }

      it 'responds to a graphql query via get' do
        get '/', {'query' => '{ hello }'}
        assert_equal 200, last_response.status
        expected_body = '{"data":{"hello":"world"}}'
        assert_equal expected_body, last_response.body
      end

      it 'responds to a graphql query via post' do
        post '/', {'query' => '{ hello }'}.to_json
        assert_equal 200, last_response.status
        expected_body = '{"data":{"hello":"world"}}'
        assert_equal expected_body, last_response.body
      end

      it 'does not support PUT queries' do
        put '/', {'query' => '{ hello }'}
        assert_equal 405, last_response.status
        assert_equal 'GraphQL Server supports only GET/POST requests', last_response.body
      end

      it 'does not support DELETE queries' do
        delete '/', {'query' => '{ hello }'}
        assert_equal 405, last_response.status
        assert_equal 'GraphQL Server supports only GET/POST requests', last_response.body
      end

      it 'returns error 400 when post body is missing' do
        post '/'
        assert_equal 400, last_response.status
        assert_equal 'POST body missing', last_response.body
      end
    end
  end

  describe "middleware" do
    describe "#call" do
      let(:app) { HelloWorldServer.middleware }

      it 'responds to a graphql query via get' do
        get '/', {'query' => '{ hello }'}
        assert_equal 200, last_response.status
        expected_body = '{"data":{"hello":"world"}}'
        assert_equal expected_body, last_response.body
      end
    end
  end
end


lib/graphql_server.rb
require 'graphql'
require 'json'
require 'rack'

class GraphQLServer
  class InvalidRequestType < Exception; end;
  class PostBodyMissing < Exception; end;

  # Initilizes GraphQLServer as a Rack app or middleware
  #
  # This Rack middleware (or app) implements a spec-compliant GraphQL server which can be queried from any GraphQL client.
  # It can be used with a provided GraphQL schema or can build one from a type definition and a resolver hash.
  #
  # @param [Array<Object>] *args The first argument should be `app` when used as a middleware
  # @param String type_def A schema definition string, or a path to a file containing the definition
  # @param Hash resolver A hash with callables for handling field resolution
  # @param GraphQL::Schema schema Use this schema if `type_def` and `resolver` is nil
  # @param Hash context
  def initialize(*args, type_def: nil, resolver: nil, schema: nil, context: nil)
    @app = args && args[0]
    @context = context
    @schema = type_def && resolver ? GraphQL::Schema.from_definition(type_def, default_resolve: resolver) : schema
  end

  def middleware?
    !@app.nil?
  end

  def call(env)
    request = Rack::Request.new(env)
    
    # graphql accepts GET and POST requests
    raise InvalidRequestType unless request.get? || request.post?

    payload = if request.get?
      request.params
    elsif request.post?
      body = request.body.read
      raise PostBodyMissing if body.empty?
      JSON.parse(body)
    end

    response = @schema.execute(
      payload['query'], 
      variables: payload['variables'],
      operation_name: payload['operationName'],
      context: @context, 
    ).to_json
    
    [200, {'Content-Type' => 'application/json', 'Content-Length' => response.bytesize.to_s}, [response]]
  rescue InvalidRequestType
    # Method Not Allowed
    [405, {"Content-Type" => "text/html"}, ["GraphQL Server supports only GET/POST requests"]]
  rescue PostBodyMissing
    # Bad Request
    [400, {"Content-Type" => "text/html"}, ["POST body missing"]]
  end
end
